### ISSUE TITLE
[Community Pipeline] Imagic: Text-Based Real Image Editing with Diffusion Models

### ISSUE THREAD

**Issue by apolinario** [huggingface/diffusers#895]
## Intro

Community Pipelines are introduced in `diffusers==0.4.0` with the idea of allowing the community to quickly add, integrate, and share their custom pipelines on top of `diffusers`. 

You can find a guide about Community Pipelines [here](https://github.com/huggingface/diffusers/issues/841). You can also find all the community examples under [`examples/community/`](https://github.com/huggingface/diffusers/tree/main/examples/community). If you have questions about the Community Pipelines feature, please head to the [parent issue](https://github.com/huggingface/diffusers/issues/841).

## Idea: Imagic: Text-based Real Image Editing with Diffusion Models

This pipeline aims to implement [this paper](https://arxiv.org/abs/2210.09276) to Stable Diffusion, allowing for real-world image editing. Example from the paper: 
![image](https://user-images.githubusercontent.com/788417/196388568-4ee45edd-e990-452c-899f-c25af32939be.png)

**Comment by Alx-AI** [comment]
Would love to see this added, notebook implementation here for reference https://github.com/justinpinkney/stable-diffusion/blob/main/notebooks/imagic.ipynb

**Comment by asofiaoliveira** [comment]
I would like to work on this

**Comment by patrickvonplaten** [comment]
Awesome @asofiaoliveira !

Feel free to open a PR and to attach it here

**Comment by MarkRich** [comment]
tried my hand at this here: https://github.com/huggingface/diffusers/pull/958 please let me know if there's any comments!

**Comment by asofiaoliveira** [comment]
I guess I'll leave it to @MarkRich then 😅

**Comment by 0xdevalias** [comment]
FYI, it looks like that PR has been merged now:

- https://github.com/huggingface/diffusers/pull/958

And the implementation is available here:

- https://github.com/huggingface/diffusers/tree/main/examples/community#imagic-stable-diffusion

Can this issue be closed now?

**Comment by njucckevin** [comment]
> FYI

Is there someone try the effect with this code? I failed to achieve the effect in the paper (for example, let a doy playing with a toy) with the Imagic Stable Diffusion.

**Comment by askerlee** [comment]
@0xdevalias I don't feel that the implementation was complete/correct. For image sampling, the text embedding should be the interpolation of an optimized embedding and the target prompt embedding.

**Comment by askerlee** [comment]
@0xdevalias It seems there is some issue with the implementation? In train(), the text embedding is optimized first. But prior to that, unet and text_encoder are set to disable BP:

        self.unet.requires_grad_(False)
        self.text_encoder.requires_grad_(False)

Does this mean there won't be valid gradients back propagated from the loss to the text embedding? 
I'm not very sure. Thanks.

**Comment by 0xdevalias** [comment]
@askerlee I don't know anything about the implementation, nor really used it. I just noticed the PR and figured I'd link it here

**Comment by ShaoTengLiu** [comment]
@njucckevin I also get wrong results.
 For example, here are the results for "A photo of a bird spreading wings":
![bird_alpha_1](https://user-images.githubusercontent.com/51499992/212677983-f746fd13-16f4-47af-85ac-fe7b2acf0a4d.png)
![bird_alpha_1_5](https://user-images.githubusercontent.com/51499992/212677994-31da0cb2-022c-46ed-8823-ca5b65fe3118.png)
![bird_alpha_2](https://user-images.githubusercontent.com/51499992/212678011-491e8b53-9d65-41d1-961c-190f57a327ce.png)

Can anyone give me some hints?

**Comment by BoyuanJiang** [comment]
> @njucckevin I also get wrong results. For example, here are the results for "A photo of a bird spreading wings": ![bird_alpha_1](https://user-images.githubusercontent.com/51499992/212677983-f746fd13-16f4-47af-85ac-fe7b2acf0a4d.png) ![bird_alpha_1_5](https://user-images.githubusercontent.com/51499992/212677994-31da0cb2-022c-46ed-8823-ca5b65fe3118.png) ![bird_alpha_2](https://user-images.githubusercontent.com/51499992/212678011-491e8b53-9d65-41d1-961c-190f57a327ce.png)
> 
> Can anyone give me some hints?

I also cannot reproduce the result in the paper

**Comment by Kathy-Peng** [comment]
I am using stable diffusion to replicate their result on imagen(non open source) and used 500 optimisation steps and 1000 fine-tuning steps. The best result come from the last picture where we have lambda coefficient of 1. Not quite sure why the change comes so late.

**Comment by 1702609** [comment]
> @njucckevin I also get wrong results. For example, here are the results for "A photo of a bird spreading wings": ![bird_alpha_1](https://user-images.githubusercontent.com/51499992/212677983-f746fd13-16f4-47af-85ac-fe7b2acf0a4d.png) ![bird_alpha_1_5](https://user-images.githubusercontent.com/51499992/212677994-31da0cb2-022c-46ed-8823-ca5b65fe3118.png) ![bird_alpha_2](https://user-images.githubusercontent.com/51499992/212678011-491e8b53-9d65-41d1-961c-190f57a327ce.png)
> 
> Can anyone give me some hints?

I am getting the exact same problem. What is the solution to this?

**Comment by 1702609** [comment]
> > FYI
> 
> Is there someone try the effect with this code? I failed to achieve the effect in the paper (for example, let a doy playing with a toy) with the Imagic Stable Diffusion.

I am also facing this problem. When I set anything below 1 for lambda coefficient, I get an image that is the same as the input image. If I change the coefficient by more than 1 then I get an image of a random bird spreading its wing.

**Comment by 1702609** [comment]
> Hey guys, here is what I have: ![Unknown](https://user-images.githubusercontent.com/55202091/216996492-cd860a4e-2e3b-49b5-8b69-0419f4eed295.png)
> 
> ![Unknown-2](https://user-images.githubusercontent.com/55202091/216996523-157e1407-abab-4db4-8480-5244c94f0100.png) ![Unknown-3](https://user-images.githubusercontent.com/55202091/216996541-f7320036-1635-46c8-acaf-498cf7807eb5.png)

Did you change anything?

**Comment by 1702609** [comment]
> Hey guys, here is what I have: ![Unknown](https://user-images.githubusercontent.com/55202091/216996492-cd860a4e-2e3b-49b5-8b69-0419f4eed295.png)
> 
> ![Unknown-2](https://user-images.githubusercontent.com/55202091/216996523-157e1407-abab-4db4-8480-5244c94f0100.png) ![Unknown-3](https://user-images.githubusercontent.com/55202091/216996541-f7320036-1635-46c8-acaf-498cf7807eb5.png)

Did you change anything? How did you get that?

**Comment by WonwoongCho** [comment]
I have been trying to run the ipynb script with StableDiffusion v2, and seems like both gradient steps (1. updating embedding and 2. finetuning network parameters) do not work. Error occurs at loss.backward() part. Seems like gradient is None for some reasons. Will share more info if I can figure it out. Please get me a hint if you have some. I hope I've made a stupid mistake..

**Comment by shwetabhardwaj44** [comment]
> I am using stable diffusion to replicate their result on imagen(non open source) and used 500 optimisation steps and 1000 fine-tuning steps. The best result come from the last picture where we have lambda coefficient of 1. Not quite sure why the change comes so late.

Hi @Kathy-Peng I am also following the same config and my code is written as below. Could you kindly confirm if there is any missing step in this. In my results, the edited image doesn't change much even at lambda coefficient = 1.1

```
model_id = "CompVis/stable-diffusion-v1-4"
pipe = DiffusionPipeline.from_pretrained(
                          model_id,
                          cache_dir=CACHE_DIR,
                          safety_checker=None,
                          use_auth_token=True,
                          custom_pipeline="imagic_stable_diffusion",
                          scheduler = DDIMScheduler(\
                                      beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear",\
                                      clip_sample=False, set_alpha_to_one=False)
                          )
pipe.to("cuda")
generator = torch.Generator("cuda").manual_seed(0)

alphas = [0.8, 0.9, 1, 1.1, 1.2]
guidance_scale = [6.5, 7.0, 7.5, 8.0, 8.5, 9.5]
# curr_prompt  is given
# curr_image is loaded

res = pipe.train(curr_prompt, image=curr_image, generator=generator)
reconstructed_image = res.images[0]

## Once the pipeline is trained, run inference with different alphas and text guidance scales.
for alpha in alphas:
        for text_guide in guidance_scale:                                                                             
                edited_image  = pipe(num_inference_steps=50, alpha=alpha, guidance_scale=text_guide).images[0]
```

**Comment by HashmatShadab** [comment]
Was anyone able to reproduce the results?

**Comment by witcherofresearch** [comment]
@HashmatShadab @njucckevin @BoyuanJiang @ShaoTengLiu @1702609 @shwetabhardwaj44 Please check out my new paper Forgedit, which is much faster than Imagic and edting results are way much better than Imagic with Stable Diffusion.  https://github.com/witcherofresearch/Forgedit/
Here is a few examples on the editing results of Forgedit. For example, for the target prompt 'A photo of a bird spreading wings.' and original image 
![bird](https://github.com/witcherofresearch/Forgedit/assets/144800993/ac0bcf5f-2e8e-418a-94ba-fe56a6fb91e3)
using the DreamBoothForgedit, we could get
![6_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=0 8_alpha=1 0_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/5b1eef6b-16a1-4f72-922b-baac3acca229)
![4_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=1 1_alpha=1 3_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/30743417-3883-4684-9f0d-5e9c1b11076c)
![9_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=0 8_alpha=1 2000000000000002_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/63ecb1c4-f5fe-497f-b5cc-e942ab2b3051)
![1_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=0 8_alpha=0 9_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/a94c2e63-c6be-442e-bb5e-06eb4e6aa1ac)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/cf24e8f5-4965-412a-925f-360abdee87e1)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/53c0d914-3f1a-4495-b0b8-8d292d242da4)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/5b77718b-df8b-41d7-b1a1-ddf26df823f3)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/77f95219-2e28-4205-9922-17b5eea12e1a)

**Comment by witcherofresearch** [comment]
@HashmatShadab @njucckevin @BoyuanJiang @ShaoTengLiu @1702609 @shwetabhardwaj44 Please check out my new paper Forgedit, which is much faster than Imagic and edting results are way much better than Imagic with Stable Diffusion.  https://github.com/witcherofresearch/Forgedit/
Here is a few examples on the editing results of Forgedit. For example, for the target prompt 'A photo of a bird spreading wings.' and original image 
![bird](https://github.com/witcherofresearch/Forgedit/assets/144800993/ac0bcf5f-2e8e-418a-94ba-fe56a6fb91e3)
using the DreamBoothForgedit, we could get
![6_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=0 8_alpha=1 0_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/5b1eef6b-16a1-4f72-922b-baac3acca229)
![4_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=1 1_alpha=1 3_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/30743417-3883-4684-9f0d-5e9c1b11076c)
![9_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=0 8_alpha=1 2000000000000002_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/63ecb1c4-f5fe-497f-b5cc-e942ab2b3051)
![1_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=0 8_alpha=0 9_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/a94c2e63-c6be-442e-bb5e-06eb4e6aa1ac)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/cf24e8f5-4965-412a-925f-360abdee87e1)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/53c0d914-3f1a-4495-b0b8-8d292d242da4)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/5b77718b-df8b-41d7-b1a1-ddf26df823f3)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/77f95219-2e28-4205-9922-17b5eea12e1a)

**Comment by witcherofresearch** [comment]
@HashmatShadab @njucckevin @BoyuanJiang @ShaoTengLiu @1702609 @shwetabhardwaj44 Please check out my new paper Forgedit, which is much faster than Imagic and edting results are way much better than Imagic with Stable Diffusion.  https://github.com/witcherofresearch/Forgedit/
Here is a few examples on the editing results of Forgedit. For example, for the target prompt 'A photo of a bird spreading wings.' and original image 
![bird](https://github.com/witcherofresearch/Forgedit/assets/144800993/ac0bcf5f-2e8e-418a-94ba-fe56a6fb91e3)
using the DreamBoothForgedit, we could get
![6_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=0 8_alpha=1 0_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/5b1eef6b-16a1-4f72-922b-baac3acca229)
![4_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=1 1_alpha=1 3_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/30743417-3883-4684-9f0d-5e9c1b11076c)
![9_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=0 8_alpha=1 2000000000000002_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/63ecb1c4-f5fe-497f-b5cc-e942ab2b3051)
![1_orig_A photo of a bird spreading wings _guidance_scale=7 5__textalpha=0 8_alpha=0 9_bird jpeg](https://github.com/witcherofresearch/Forgedit/assets/144800993/a94c2e63-c6be-442e-bb5e-06eb4e6aa1ac)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/cf24e8f5-4965-412a-925f-360abdee87e1)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/53c0d914-3f1a-4495-b0b8-8d292d242da4)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/5b77718b-df8b-41d7-b1a1-ddf26df823f3)
![image](https://github.com/witcherofresearch/Forgedit/assets/144800993/77f95219-2e28-4205-9922-17b5eea12e1a)
